get_goes_x.py

```diff
@@ -11,99 +11,153 @@
+    row for each sample time, include the historical time
+    columns, and add a column to describe the satellite.
+
+    If optional satellite arg is supplied, filter the source data to include
+    only records that match that satellite.
+    """
+
+    if len(dat) == 0:
+        print('Warning: No data in fetched file')
+        sys.exit(0)
+
+    # Select only the GOES satellite specified from the args
+    if satellite is not None:
+        dat = dat[dat['satellite'] == satellite]
+        if len(dat) == 0:
+            print('Warning: No data in fetched file for satellite {}'.format(satellite))
+            sys.exit(0)
+
+    # Make a table for each of the two wavelengths
+    shortdat = dat[dat['energy'] == '0.05-0.4nm']['flux', 'satellite', 'time_tag']
+    shortdat.rename_column('flux', 'short')
+    longdat = dat[dat['energy'] == '0.1-0.8nm']['flux', 'satellite', 'time_tag']
+    longdat.rename_column('flux', 'long')
+    if len(longdat) != len(shortdat):
+        print('Warning: "short" and "long" table have mismatched lengths')
+
+    # Join them on time and satellite (seems to be OK for these data)
+    joindat = join(shortdat, longdat)
+
+    # Add a time column with Chandra secs and remove original time_tags
+    times = Time(joindat['time_tag'])
+    joindat['time'] = times.cxcsec
+    joindat.remove_column('time_tag')
+
+    # Add the other columns the old file format wanted
+    joindat['mjd'] = times.mjd.astype(int)
+    joindat['secs'] = np.array(np.round((times.mjd - joindat['mjd']) * 86400,
+                                        decimals=0)).astype(int)
+    joindat['year'] = [t.year for t in times.datetime]
+    joindat['month'] = [t.month for t in times.datetime]
+    joindat['dom'] = [t.day for t in times.datetime]
+    joindat['hhmm'] = np.array([
+        f"{t.hour}{t.minute:02}" for t in times.datetime]).astype(int)
+
+    joindat['ratio'] = -100000.0
+    ok = (joindat['long'] != 0) & (joindat['long'] != -100000.0)
+    joindat['ratio'][ok] = joindat['short'][ok] / joindat['long'][ok]
+
+    return joindat['year', 'month', 'dom', 'hhmm', 'mjd', 'secs',
+                   'short', 'long', 'ratio', 'time', 'satellite'].as_array()
+
+
+def main():
+
+    args = get_options()
+
+    # Read the data file just to get the last record
     try:
-        table = h5.root.data
-        lasttime = table.col('time')[-1]
-        ok = newdat['time'] > lasttime
-        newdat = newdat[ok]
-        h5.root.data.append(newdat)
-    except tables.NoSuchNodeError:
-        table = h5.create_table(h5.root, 'data', newdat,
-                               "GOES_X rates", expectedrows=2e7)
-    h5.root.data.flush()
-
+        with tables.open_file(args.h5, mode='r',
+                              filters=tables.Filters(complevel=5, complib='zlib')) as h5:
+            table = h5.root.data
+            lasttime = table.col('time')[-1]
+    except (OSError, IOError, tables.NoSuchNodeError):
```

Comment:
Is it too hideous to wrap the whole chunk in a try and handle missing file or node?

Reply:
> wrap the whole chunk in a try and handle missing file or node?

I don't understand this idea, starting with what "the whole chunk" is.  I assume that you have tested this code for a non-existent file where it ends up correctly creating the file from scratch.

Reply:
IIRC from 23 days ago when I made the comment, your original feedback on the original code was to add IOError handling on opening the file in append mode, and at that time there was a separate try/except with NoSuchNodeError on the lines that try to address h5.root.data .  Here I've just put the open_file, the access of h5.root.data and the read of 'time' all in the same try/except, which I'm calling "the whole chunk".  I also assume I tested this code for these various error conditions but I suppose since I didn't actually note that in the thread 23 days ago that I should go and do it again.

Reply:
Ah, so you are asking if the current code at this point is OK.  Sure, looks fine to me.  I thought you were asking about yet a different strategy.